fendermint/app/options/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::path::PathBuf;
+
+use clap::{Args, Parser, Subcommand, ValueEnum};
+use fvm_shared::address::Network;
+
+use self::{eth::EthArgs, genesis::GenesisArgs, key::KeyArgs, rpc::RpcArgs, run::RunArgs};
+
+pub mod eth;
+pub mod genesis;
+pub mod key;
+pub mod rpc;
+pub mod run;
+
+mod parse;
+
+use parse::parse_network;
+
+/// Parse the main arguments by:
+/// 1. Parsing the [GlobalOptions]
+/// 2. Setting any system wide parameters based on the globals
+/// 3. Parsing and returning the final [Options]
+pub fn parse() -> Options {
+    let opts: GlobalOptions = GlobalOptions::parse();
+    fvm_shared::address::set_current_network(opts.global.network);
+    let opts: Options = Options::parse();
+    opts
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
+pub enum LogLevel {
+    Off,
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+/// Options which set some global value that affects the parsing of other options.
+#[derive(Args, Debug)]
+pub struct GlobalArgs {
+    /// Set the FVM Network.
+    #[arg(short, long, default_value = "mainnet", env = "FVM_NETWORK", value_parser = parse_network)]
+    pub network: Network,
+}
+
+/// A version of options that does partial matching on the arguments, with its only interest
+/// being the capture of global parameters that need to take effect first, before we parse [Options].
+///
+/// This one doesn't handle `--help` or `help` so that it is passed on to the next parser,
+/// where the full set of commands and arguments can be printed properly.
+#[derive(Parser, Debug)]
+#[command(version, disable_help_flag = true)]
+pub struct GlobalOptions {
+    #[command(flatten)]
+    pub global: GlobalArgs,
+
+    /// Capture all the normal commands, basically to ingore them.
+    #[arg(allow_hyphen_values = true, trailing_var_arg = true)]
+    pub cmd: Vec<String>,
+}
+
+#[derive(Parser, Debug)]
+#[command(version)]
+pub struct Options {
+    /// Set a custom directory for data and configuration files.
+    #[arg(
+        short = 'd',
+        long,
+        default_value = "~/.fendermint",
+        env = "FM_HOME_DIR"
+    )]
+    pub home_dir: PathBuf,
+
+    /// Optionally override the default configuration.
+    #[arg(short, long, default_value = "dev")]
+    pub mode: String,
+
+    /// Set the logging level.
+    #[arg(short, long, default_value = "info", value_enum, env = "LOG_LEVEL")]
+    pub log_level: LogLevel,
+
+    /// Global options repeated here for discoverability.
+    #[command(flatten)]
+    pub global: GlobalArgs,
```

Comment:
what does it mean by `discoverability`? Do we have to config this parameter?

Reply:
I'll change it to `FM_NETWORK`, maybe that's more what people would expect.